### PR TITLE
Config version support & migration refactor

### PR DESCRIPTION
**Pull request by gwicke**

### CODE REVIEW

**Comment by gwicke** [comment]
Rebased on top of #168, so ready to go.

**Comment by gwicke** [comment]
@d00rman @eevans @Pchelolo, could you have a look?

**Comment by Pchelolo** [comment]
Looks like it's still not on master, needs yet another rebase.

File: lib/schemaMigration.js [hunk 1]
```diff
@@ -242,117 +246,165 @@ SecondaryIndexes.prototype.migrate = function() {
...
+ * @throws  {Error} if the proposed migration fails to validate
  * @return a promise that resolves when the migration tasks are complete
  */
-SchemaMigrator.prototype.migrate = function() {
-    return P.each(this.migrators, function(migrator) {
-        return migrator.migrate();
+SchemaMigrator.prototype.migrate = function(req, current, proposed) {
+    var toMigrate = [];
+    // First phase: validate everything.
+    this.migrators.forEach(function(migrator) {
```
Comment (Pchelolo) [inline lib/schemaMigration.js]: Nit: may be done cleaner with `filter` method

Comment (gwicke) [inline lib/schemaMigration.js]: Yup, good point.

**Pull request merged**